File: Dmoj/DMOPC '15 Contest 3/P3 - Dimethylbenzene.cpp
// https://dmoj.ca/problem/dmopc15c3p3
// DMOPC '15 Contest 3 P3 - Dimethylbenzene


#include <iostream>
#include <queue>
#include <vector>

#define pii pair<int, int>

using namespace std;

int n, m, x, y, node, dis;
vector<int> adje[100001];
queue<pii> que;
bool visited[100001];

int main() {
    cin >> n >> m;
    memset(visited, false, sizeof(visited));
    for (int i = 0; i < m; i++) {
        cin >> x >> y;
        adje[x].push_back(y);
    }

    que.push(make_pair(1, 0));

    while (!que.empty()) {
        node = que.front().first;
        dis = que.front().second;
        que.pop();

        if (node == 1 && dis >= 6) {
            cout << "YES" << endl;
            return 0;
        }

        for (auto i: adje[node]) {
            if (!visited[i]) {
                que.push(make_pair(i, dis+1));
                visited[i] = true;
            }
        }
    }

    cout << "NO" << endl;
    return 0;
}

File: Dmoj/DMOPC '15 Contest 3/P3 - Dimethylbenzene.py
#  https://dmoj.ca/problem/dmopc15c3p3
#  DMOPC '15 Contest 3 P3 - Dimethylbenzene

from collections import deque
def solve():
    n, m = map(int, input().split())

    visited = [False for i in range(n+1)]

    adj = [[] for i in range(n+1)]

    for i in range(m):
        x, y = map(int, input().split())
        adj[x].append(y)

    dq = deque()

    dq.append((1, 0))

    while len(dq) > 0:
        node, dis = dq.pop()
        if node == 1 and dis >= 6:
            print("YES")
            return 0
        
        for i in adj[node]:
            if not visited[i]:
                dq.append((i, dis+1))
                visited[i] = True
        
    print("NO")

solve()
